src/main/cpp/win_fsnotifier.cpp

```diff
@@ -51,99 +176,96 @@ void handlePathChanged(watch_details_t *details, FILE_NOTIFY_INFORMATION *info)
-        return 0;
+        return;
     }
 
-    OVERLAPPED overlapped;
-    memset(&overlapped, 0, sizeof(overlapped));
-    overlapped.hEvent = CreateEvent(NULL, FALSE, FALSE, NULL);
-
-    HANDLE hDrive = CreateFileW(details->drivePath, GENERIC_READ, CREATE_SHARE, NULL, OPEN_EXISTING, CREATE_FLAGS, NULL);
-
-    char buffer[EVENT_BUFFER_SIZE];
-    HANDLE handles[2] = {details->stopEventHandle, overlapped.hEvent};
-    while (true) {
-        int rcDrive = ReadDirectoryChangesW(hDrive, buffer, sizeof(buffer), TRUE, EVENT_MASK, NULL, &overlapped, NULL);
-        if (rcDrive == 0) {
-            // TODO Error handling
-            printf("~~~~ Couldn't read directory: %d - %d\n", rcDrive, GetLastError());
-            break;
-        }
-
-        DWORD rc = WaitForMultipleObjects(2, handles, FALSE, INFINITE);
-        if (rc == WAIT_OBJECT_0) {
-            break;
-        }
-        if (rc == WAIT_OBJECT_0 + 1) {
-            DWORD dwBytesReturned;
-            if (!GetOverlappedResult(hDrive, &overlapped, &dwBytesReturned, FALSE)) {
-                // TODO Error handling
-                printf("~~~~ Failed to wait\n");
-                break;
-            }
-
-            if (dwBytesReturned == 0) {
-                // don't send dirty too much, everything is changed anyway
-                // TODO Understand what this does
-                if (WaitForSingleObject(details->stopEventHandle, 500) == WAIT_OBJECT_0)
-                    break;
-
-                // Got a buffer overflow => current changes lost => send INVALIDATE on root
-                reportEvent(FILE_EVENT_INVALIDATE, details->drivePath, 3, details);
-            } else {
-                FILE_NOTIFY_INFORMATION *info = (FILE_NOTIFY_INFORMATION *)buffer;
-                do {
-                    handlePathChanged(details, info);
-                    info = (FILE_NOTIFY_INFORMATION *)((char *)info + info->NextEntryOffset);
-                } while (info->NextEntryOffset != 0);
-            }
-        }
+    while (!terminate || watchPoints.size() > 0) {
+        SleepEx(INFINITE, true);
     }
 
     printf("~~~~ Stopping thread\n");
 
-    CloseHandle(overlapped.hEvent);
-    CloseHandle(hDrive);
-
     // TODO Extract this logic to some shared function
     jint statDetach = jvm->DetachCurrentThread();
     if (statDetach != JNI_OK) {
         printf("Failed to detach JNI from current thread: %d\n", statAttach);
-        return 0;
+        return;
     }
+}
 
-    return 0;
+static void CALLBACK requestTerminationCallback(_In_ ULONG_PTR arg) {
+    Server* server = (Server*)arg;
+    server->requestTermination();
+}
+
+void Server::requestTermination() {
+    terminate = true;
+    for (auto &watchPoint : watchPoints) {
+        watchPoint->close();
+    }
+}
+
+void Server::close(JNIEnv *env) {
+    QueueUserAPC(requestTerminationCallback, this->threadHandle, (ULONG_PTR)this);
+    WaitForSingleObject(this->threadHandle, INFINITE);
+    CloseHandle(this->threadHandle);
```

Comment:
nit: Very little developers check the return code of `CloseHandle`, however, according to the documentation, it can fail.

Reply:
It's more about what to do if it fails. What's the course of action recommended then?